### ISSUE TITLE
Multi-hop Attention Graph Neural Networks

### ISSUE THREAD

**Issue by InabaTatsuro** [InabaTatsuro/papers#14]
# Multi-hop Attention Graph Neural Networks [[Wang+, IJCAI21](https://www.ijcai.org/proceedings/2021/425)]
![スクリーンショット 2023-10-17 20 32 17](https://github.com/InabaTatsuro/papers/assets/102784221/016d258d-4af2-4951-80c5-4b69bd2bec8e)

the left is previous (like GAT[#4]), the right is MAGNA

## Abstract
- Multi-hop Attention Graph Neural Network (MAGNA) first computes attention weights on edges
- then, computes self-attention weight between disconnected pairs of nodes using the attention weights on the edges
- MAGNA can capture long-range interactions and compute attention between non-connected nodes

### Edge Attention Computation
![スクリーンショット 2023-10-17 20 51 39](https://github.com/InabaTatsuro/papers/assets/102784221/49e3c8b4-8e98-4f50-b1f2-6fdfd6f26abd)

- concat 3 embeddings, linear transformations of 2 node embeddings and 1 edge embeddings
- then, linear transform into a value and apply LeakyReLU to get attention score matrix
- obtain the attention matrix by performing row-wise softmax over the score matrix

### Attention Diffusion for Multi-hop Neighbors
<img src="https://github.com/InabaTatsuro/papers/assets/102784221/fed86925-dd27-4283-9553-448093059ff5" width="400">
<br>
<img src="https://github.com/InabaTatsuro/papers/assets/102784221/3313ae51-c08a-4efd-afd7-16f24e72c243" width="400">

### MAGNA architecture
![スクリーンショット 2023-10-17 23 00 56](https://github.com/InabaTatsuro/papers/assets/102784221/72e42c85-77fb-4803-ab63-a8ae23f097e7)

- multi-head

<img width="400" alt="スクリーンショット 2023-10-18 14 37 21" src="https://github.com/InabaTatsuro/papers/assets/102784221/a0d7024f-1e93-4cc6-bc8a-7ddc3a9f6899">

- The above formula represents GAT. The differences with GAT are
  - (1) MAGNA removes the restriction of attending to only direct neighbors by using $A$ induced from A
  - (2) MAGNA uses layer normalization and MLP layers to achieve higher expression
